File: generate_items_inserts.js
const fs = require('fs');
const path = require('path');

function cleanText(text) {
    if (typeof text !== 'string') {
        return text;
    }
    // Remove {@tag text|source|...} and similar patterns
    text = text.replace(/{@\w+\s*([^|]+?)(?:\|[^}]+?)?}/g, '$1');
    // Escape single quotes by doubling them for SQL
    text = text.replace(/'/g, "''");
    // No need to escape double quotes here, JSON.stringify will handle it.
    return text;
}

function parseItemEntries(itemId, entries, dbType) {
    const insertStatements = [];
    if (!entries || !Array.isArray(entries)) {
        return insertStatements;
    }

    for (let i = 0; i < entries.length; i++) {
        const entry = entries[i];
        const entryType = entry.type || 'string';
        let content = null;
        let data = null;

        if (entryType === 'string') {
            content = cleanText(entry);
        } else {
            // For complex types, stringify the whole object to JSON
            data = JSON.stringify(entry);
            // Escape single quotes within the JSON string for SQL literal
            if (data) {
                data = data.replace(/'/g, "''");
            }
        }

        if (dbType === 'postgres') {
            insertStatements.push(
                `INSERT INTO item_entries (item_id, entry_order, type, content, data) VALUES (` +
                `${itemId}, ${i}, '${entryType}', ${content !== null ? "'" + content + "'" : 'NULL'}, ` +
                `${data !== null ? "'" + data + "'::jsonb" : 'NULL'});`
            );
        } else if (dbType === 'sqlite') {
            insertStatements.push(
                `INSERT INTO item_entries (item_id, entry_order, type, content, data) VALUES (` +
                `${itemId}, ${i}, '${entryType}', ${content !== null ? "'" + content + "'" : 'NULL'}, ` +
                `${data !== null ? "'" + data + "'" : 'NULL'});`
            );
        }
    }
    return insertStatements;
}

function generateInserts(jsonData, dbType, existingItemNames, existingSourceNames) {
    const itemInserts = [];
    const itemEntryInserts = [];
    const sourceInserts = [];
    const sourcesMap = {}; // To map source names to their IDs for this run

    // Populate sourcesMap with existing sources to maintain consistent IDs
    // This is a simplification. In a real scenario, you'd query the DB for max IDs.
    // For script generation, we'll just ensure new sources get new IDs.
    let currentSourceId = 1;
    for (const sourceName of existingSourceNames) {
        sourcesMap[sourceName] = currentSourceId++;
    }

    for (const itemData of jsonData) {
        const name = cleanText(itemData.name);

        // Skip if item already exists
        if (existingItemNames.has(name)) {
            continue;
        }

        const source = cleanText(itemData.source);
        const page = itemData.page !== undefined && itemData.page !== null ? itemData.page : 'NULL';
        const itemType = itemData.type !== undefined && itemData.type !== null ? cleanText(itemData.type) : 'NULL';
        const rarity = itemData.rarity !== undefined && itemData.rarity !== null ? cleanText(itemData.rarity) : 'NULL';
        const weight = itemData.weight !== undefined && itemData.weight !== null ? itemData.weight : 'NULL';
        const value = itemData.value !== undefined && itemData.value !== null ? itemData.value : 'NULL';
        const attunement = itemData.reqAttune !== undefined && itemData.reqAttune !== null;
        const reqAttune = itemData.reqAttune !== undefined && itemData.reqAttune !== null ? cleanText(itemData.reqAttune) : 'NULL';
        const tier = itemData.tier !== undefined && itemData.tier !== null ? cleanText(itemData.tier) : 'NULL';
        const recharge = itemData.recharge !== undefined && itemData.recharge !== null ? cleanText(itemData.recharge) : 'NULL';
        const charges = itemData.charges !== undefined && itemData.charges !== null ? itemData.charges : 'NULL';
        const dmg1 = itemData.dmg1 !== undefined && itemData.dmg1 !== null ? cleanText(itemData.dmg1) : 'NULL';
        const dmg2 = itemData.dmg2 !== undefined && itemData.dmg2 !== null ? cleanText(itemData.dmg2) : 'NULL';
        const dmgType = itemData.dmgType !== undefined && itemData.dmgType !== null ? cleanText(itemData.dmgType) : 'NULL';
        const rangeVal = itemData.range !== undefined && itemData.range !== null ? cleanText(itemData.range) : 'NULL';
        const weaponCategory = itemData.weaponCategory !== undefined && itemData.weaponCategory !== null ? cleanText(itemData.weaponCategory) : 'NULL';
        const ammoType = itemData.ammoType !== undefined && itemData.ammoType !== null ? cleanText(itemData.ammoType) : 'NULL';

        // Booleans
        const isContainer = !!itemData.container;
        const isStaff = !!itemData.staff;
        const isWand = !!itemData.wand;
        const isRod = !!itemData.rod;
        const isSentient = !!itemData.sentient;
        const isCursed = !!itemData.curse;
        const isSrd = !!itemData.srd;
        const isBasicRules = !!itemData.basicRules;

        // Handle properties (array of strings)
        const properties = itemData.property;
        let propertiesJson = 'NULL';
        if (properties) {
            propertiesJson = JSON.stringify(properties);
            propertiesJson = propertiesJson.replace(/'/g, "''"); // Escape single quotes within JSON string
        }

        // Handle roll_data (can be complex, store as JSON)
        const rollData = itemData.roll;
        let rollDataJson = 'NULL';
        if (rollData) {
            rollDataJson = JSON.stringify(rollData);
            rollDataJson = rollDataJson.replace(/'/g, "''"); // Escape single quotes within JSON string
        }

        // Handle misc_data (catch-all for other fields not explicitly mapped)
        const miscDataKeys = [
            "name", "source", "page", "type", "rarity", "weight", "value",
            "reqAttune", "tier", "recharge", "charges", "roll", "dmg1", "dmg2",
            "dmgType", "range", "weaponCategory", "ammoType", "container",
            "staff", "wand", "rod", "sentient", "curse", "srd", "basicRules",
            "property", "entries", "_meta", "reqAttuneTags", "bonusSpellAttack",
            "bonusSpellSaveDc", "focus", "bonusWeapon", "hasFluffImages",
            "hasFluff", "resist", "conditionImmune", "ability", "attachedSpells",
            "group", "scfType", "detail1", "tattoo", "hasRefs", "additionalSources",
            "additionalEntries", "lootTables", "miscTags", "age", "arrow",
            "packContents", "needleBlowgun", "bolt", "dagger", "bulletFirearm",
            "bulletSling", "club", "crossbow", "hammer", "bow", "sword", "polearm",
            "lance", "net", "spear", "rapier", "scfType", "ac", "strength", "stealth",
            "vehAc", "vehHp", "vehSpeed", "crew", "capPassenger", "capCargo",
            "cellEnergy", "light", "modifySpeed"
        ];
        const miscData = {};
        for (const key in itemData) {
            if (itemData.hasOwnProperty(key) && !miscDataKeys.includes(key)) {
                miscData[key] = itemData[key];
            }
        }
        let miscDataJson = 'NULL';
        if (Object.keys(miscData).length > 0) {
            miscDataJson = JSON.stringify(miscData);
            miscDataJson = miscDataJson.replace(/'/g, "''"); // Escape single quotes within JSON string
        }

        // Determine stealth_disadvantage and strength_req for armor
        let stealthDisadvantage = false;
        let strengthReq = 'NULL';
        let armorAc = 'NULL';
        let armorDexMod = false;

        if (['LA', 'MA', 'HA'].includes(itemType)) { // Light, Medium, Heavy Armor
            armorAc = itemData.ac !== undefined && itemData.ac !== null ? itemData.ac : 'NULL';
            if (itemData.stealth === true) {
                stealthDisadvantage = true;
            }
            if (itemData.strength !== undefined && itemData.strength !== null) {
                try {
                    strengthReq = parseInt(itemData.strength);
                } catch (e) {
                    strengthReq = 'NULL';
                }
            }

            if (itemType === 'LA') { // Light Armor
                armorDexMod = true;
            } else if (itemType === 'MA') { // Medium Armor
                armorDexMod = true;
            } else if (itemType === 'HA') { // Heavy Armor
                armorDexMod = false;
            }
        }

        // Add source to the sourcesMap if not already present
        if (!(source in sourcesMap)) {
            sourcesMap[source] = currentSourceId++;
            sourceInserts.push(`INSERT INTO sources (name) VALUES ('${source}');`);
        }
        
        const sourceId = sourcesMap[source];

        // Construct the INSERT statement for the items table
        if (dbType === 'postgres') {
            itemInserts.push(
                `INSERT INTO items (name, source_id, page, type, rarity, weight, value, attunement, req_attune, tier, recharge, charges, roll_data, dmg1, dmg2, dmg_type, properties, range, stealth_disadvantage, strength_req, armor_ac, armor_dex_mod, weapon_category, ammo_type, is_container, is_staff, is_wand, is_rod, is_sentient, is_cursed, is_srd, is_basic_rules, misc_data) VALUES (` +
                `'${name}', ${sourceId}, ${page}, ${itemType !== 'NULL' ? "'" + itemType + "'" : 'NULL'}, ${rarity !== 'NULL' ? "'" + rarity + "'" : 'NULL'}, ${weight}, ${value}, ${attunement}, ${reqAttune !== 'NULL' ? "'" + reqAttune + "'" : 'NULL'}, ${tier !== 'NULL' ? "'" + tier + "'" : 'NULL'}, ${recharge !== 'NULL' ? "'" + recharge + "'" : 'NULL'}, ${charges}, ${rollDataJson !== 'NULL' ? "'" + rollDataJson + "'::jsonb" : 'NULL'}, ${dmg1 !== 'NULL' ? "'" + dmg1 + "'" : 'NULL'}, ${dmg2 !== 'NULL' ? "'" + dmg2 + "'" : 'NULL'}, ${dmgType !== 'NULL' ? "'" + dmgType + "'" : 'NULL'}, ${propertiesJson !== 'NULL' ? "'" + propertiesJson + "'::jsonb" : 'NULL'}, ${rangeVal !== 'NULL' ? "'" + rangeVal + "'" : 'NULL'}, ${stealthDisadvantage}, ${strengthReq}, ${armorAc}, ${armorDexMod}, ${weaponCategory !== 'NULL' ? "'" + weaponCategory + "'" : 'NULL'}, ${ammoType !== 'NULL' ? "'" + ammoType + "'" : 'NULL'}, ${isContainer}, ${isStaff}, ${isWand}, ${isRod}, ${isSentient}, ${isCursed}, ${isSrd}, ${isBasicRules}, ${miscDataJson !== 'NULL' ? "'" + miscDataJson + "'::jsonb" : 'NULL'});`
            );
        } else if (dbType === 'sqlite') {
            itemInserts.push(
                `INSERT INTO items (name, source_id, page, type, rarity, weight, value, attunement, req_attune, tier, recharge, charges, roll_data, dmg1, dmg2, dmg_type, properties, range, stealth_disadvantage, strength_req, armor_ac, armor_dex_mod, weapon_category, ammo_type, is_container, is_staff, is_wand, is_rod, is_sentient, is_cursed, is_srd, is_basic_rules, misc_data) VALUES (` +
                `'${name}', ${sourceId}, ${page}, ${itemType !== 'NULL' ? "'" + itemType + "'" : 'NULL'}, ${rarity !== 'NULL' ? "'" + rarity + "'" : 'NULL'}, ${weight}, ${value}, ${attunement ? 1 : 0}, ${reqAttune !== 'NULL' ? "'" + reqAttune + "'" : 'NULL'}, ${tier !== 'NULL' ? "'" + tier + "'" : 'NULL'}, ${recharge !== 'NULL' ? "'" + recharge + "'" : 'NULL'}, ${charges}, ${rollDataJson !== 'NULL' ? "'" + rollDataJson + "'" : 'NULL'}, ${dmg1 !== 'NULL' ? "'" + dmg1 + "'" : 'NULL'}, ${dmg2 !== 'NULL' ? "'" + dmg2 + "'" : 'NULL'}, ${dmgType !== 'NULL' ? "'" + dmgType + "'" : 'NULL'}, ${propertiesJson !== 'NULL' ? "'" + propertiesJson + "'" : 'NULL'}, ${rangeVal !== 'NULL' ? "'" + rangeVal + "'" : 'NULL'}, ${stealthDisadvantage ? 1 : 0}, ${strengthReq}, ${armorAc}, ${armorDexMod ? 1 : 0}, ${weaponCategory !== 'NULL' ? "'" + weaponCategory + "'" : 'NULL'}, ${ammoType !== 'NULL' ? "'" + ammoType + "'" : 'NULL'}, ${isContainer ? 1 : 0}, ${isStaff ? 1 : 0}, ${isWand ? 1 : 0}, ${isRod ? 1 : 0}, ${isSentient ? 1 : 0}, ${isCursed ? 1 : 0}, ${isSrd ? 1 : 0}, ${isBasicRules ? 1 : 0}, ${miscDataJson !== 'NULL' ? "'" + miscDataJson + "'" : 'NULL'});`
            );
        }

        const itemIdPlaceholder = itemInserts.length; // Pseudo-ID for item_entries
        const entries = itemData.entries;
        if (entries) {
            itemEntryInserts.push(...parseItemEntries(itemIdPlaceholder, entries, dbType));
        }
    }

    return { sourceInserts, itemInserts, itemEntryInserts };
}

function getExistingItemAndSourceNames(filePath) {
    const existingItemNames = new Set();
    const existingSourceNames = new Set();
    if (fs.existsSync(filePath)) {
        const content = fs.readFileSync(filePath, 'utf-8');
        const itemRegex = /INSERT INTO items \(name,.*?VALUES \('(.*?)',/g;
        const sourceRegex = /INSERT INTO sources \(name\) VALUES \('(.*?)'\);/g;

        let match;
        while ((match = itemRegex.exec(content)) !== null) {
            existingItemNames.add(match[1].replace(/''/g, "'")); // Unescape single quotes
        }
        while ((match = sourceRegex.exec(content)) !== null) {
            existingSourceNames.add(match[1].replace(/''/g, "'")); // Unescape single quotes
        }
    }
    return { existingItemNames, existingSourceNames };
}

async function main() {
    const scriptDir = 'C:\\Projects\\dnd-class-data';
    const rawDataDir = path.join(scriptDir, 'raw-data');
    const postgresExistingInsertsPath = path.join(scriptDir, 'postgres', 'insert_items_postgres.sql');
    const sqliteExistingInsertsPath = path.join(scriptDir, 'sqlite', 'insert_items_sqlite.sql');

    const itemsJsonPath = path.join(rawDataDir, 'items.json');
    const itemsBaseJsonPath = path.join(rawDataDir, 'items-base.json');

    let allItemsData = [];

    try {
        const itemsData = JSON.parse(fs.readFileSync(itemsJsonPath, 'utf-8'));
        allItemsData.push(...(itemsData.item || []));
    } catch (error) {
        console.error(`Error reading items.json: ${error.message}`);
    }

    try {
        const itemsBaseData = JSON.parse(fs.readFileSync(itemsBaseJsonPath, 'utf-8'));
        allItemsData.push(...(itemsBaseData.baseitem || []));
    } catch (error) {
        console.error(`Error reading items-base.json: ${error.message}`);
    }

    // Get existing item and source names for de-duplication
    const { existingItemNames: pgExistingItemNames, existingSourceNames: pgExistingSourceNames } = getExistingItemAndSourceNames(postgresExistingInsertsPath);
    const { existingItemNames: sqliteExistingItemNames, existingSourceNames: sqliteExistingSourceNames } = getExistingItemAndSourceNames(sqliteExistingInsertsPath);

    // Generate Postgres inserts
    const { sourceInserts: pgSourceInserts, itemInserts: pgItemInserts, itemEntryInserts: pgItemEntryInserts } = generateInserts(allItemsData, 'postgres', pgExistingItemNames, pgExistingSourceNames);
    fs.writeFileSync(path.join(scriptDir, 'insert_items_postgres.sql'), `-- Source Inserts\n` + pgSourceInserts.join('\n') + `\n\n-- Item Inserts\n` + pgItemInserts.join('\n') + `\n\n-- Item Entry Inserts\n` + pgItemEntryInserts.join('\n'), 'utf-8');

    // Generate SQLite inserts
    const { sourceInserts: sqliteSourceInserts, itemInserts: sqliteItemInserts, itemEntryInserts: sqliteItemEntryInserts } = generateInserts(allItemsData, 'sqlite', sqliteExistingItemNames, sqliteExistingSourceNames);
    fs.writeFileSync(path.join(scriptDir, 'insert_items_sqlite.sql'), `-- Source Inserts\n` + sqliteSourceInserts.join('\n') + `\n\n-- Item Inserts\n` + sqliteItemInserts.join('\n') + `\n\n-- Item Entry Inserts\n` + sqliteItemEntryInserts.join('\n'), 'utf-8');

    console.log('Generated insert_items_postgres.sql and insert_items_sqlite.sql in the root directory.');
}

main();

File: generate_items_inserts.py

import json
import os

def clean_text(text):
    """Removes markdown-like markers from text."""
    if not isinstance(text, str):
        return text
    # Remove {@tag text|source|...} and similar patterns
    text = text.replace('"', '"')
    text = text.replace("'", "''")
    return text

def parse_item_entries(item_id, entries, db_type):
    insert_statements = []
    for i, entry in enumerate(entries):
        entry_type = entry.get("type", "string")
        content = None
        data = None

        if entry_type == "string":
            content = clean_text(entry.get('data')) # Assuming 'data' key holds the string content
        elif entry_type == "list":
            # For lists, store the whole structure as JSON
            data = json.dumps(entry.get('data'))
        elif entry_type == "entries":
            # For nested entries, store the whole structure as JSON
            data = json.dumps(entry.get('data'))
        elif entry_type == "table":
            # For tables, store the whole structure as JSON
            data = json.dumps(entry.get('data'))
        else:
            # Default to string if type is unknown or not specified
            content = clean_text(str(entry.get('data')))

        if db_type == "postgres":
            insert_statements.append(
                f"INSERT INTO item_entries (item_id, entry_order, type, content, data) VALUES (" 
                f"{item_id}, {i}, '{entry_type}', {f"'{content}'" if content is not None else "NULL"}, " 
                f"{f"'{data}'::jsonb" if data is not None else "NULL"});"
            )
        elif db_type == "sqlite":
            insert_statements.append(
                f"INSERT INTO item_entries (item_id, entry_order, type, content, data) VALUES (" 
                f"{item_id}, {i}, '{entry_type}', {f"'{content}'" if content is not None else "NULL"}, " 
                f"{f"'{data}'" if data is not None else "NULL"});"
            )
    return insert_statements

def generate_inserts(json_data, db_type):
    item_inserts = []
    item_entry_inserts = []
    source_inserts = []
    sources = {}  # To store unique sources and their IDs

    # Start source ID from 1 for both databases
    current_source_id = 1

    for item_data in json_data:
        name = clean_text(item_data.get("name"))
        source = clean_text(item_data.get("source"))
        page = item_data.get("page")
        item_type = clean_text(item_data.get("type"))
        rarity = clean_text(item_data.get("rarity"))
        weight = item_data.get("weight")
        value = item_data.get("value")
        attunement = item_data.get("reqAttune") is not None
        req_attune = clean_text(item_data.get("reqAttune"))
        tier = clean_text(item_data.get("tier"))
        recharge = clean_text(item_data.get("recharge"))
        charges = item_data.get("charges")
        dmg1 = clean_text(item_data.get("dmg1"))
        dmg2 = clean_text(item_data.get("dmg2"))
        dmg_type = clean_text(item_data.get("dmgType"))
        range_val = clean_text(item_data.get("range"))
        weapon_category = clean_text(item_data.get("weaponCategory"))
        ammo_type = clean_text(item_data.get("ammoType"))

        # Booleans
        is_container = bool(item_data.get("container"))
        is_staff = bool(item_data.get("staff"))
        is_wand = bool(item_data.get("wand"))
        is_rod = bool(item_data.get("rod"))
        is_sentient = bool(item_data.get("sentient"))
        is_cursed = bool(item_data.get("curse"))
        is_srd = bool(item_data.get("srd"))
        is_basic_rules = bool(item_data.get("basicRules"))

        # Handle properties (array of strings)
        properties = item_data.get("property")
        properties_json = json.dumps(properties) if properties else None

        # Handle roll_data (can be complex, store as JSON)
        roll_data = item_data.get("roll")
        roll_data_json = json.dumps(roll_data) if roll_data else None

        # Handle misc_data (catch-all for other fields not explicitly mapped)
        misc_data = {k: v for k, v in item_data.items() if k not in [
            "name", "source", "page", "type", "rarity", "weight", "value",
            "reqAttune", "tier", "recharge", "charges", "roll", "dmg1", "dmg2",
            "dmgType", "range", "weaponCategory", "ammoType", "container",
            "staff", "wand", "rod", "sentient", "curse", "srd", "basicRules",
            "property", "entries", "_meta", "reqAttuneTags", "bonusSpellAttack",
            "bonusSpellSaveDc", "focus", "bonusWeapon", "hasFluffImages",
            "hasFluff", "resist", "conditionImmune", "ability", "attachedSpells",
            "group", "scfType", "detail1", "tattoo", "hasRefs", "additionalSources",
            "additionalEntries", "lootTables", "miscTags", "age", "arrow",
            "packContents", "needleBlowgun", "bolt", "dagger", "bulletFirearm",
            "bulletSling", "club", "crossbow", "hammer", "bow", "sword", "polearm",
            "lance", "net", "spear", "rapier", "scfType", "ac", "strength", "stealth",
            "vehAc", "vehHp", "vehSpeed", "crew", "capPassenger", "capCargo",
            "cellEnergy", "light", "modifySpeed"
        ]}
        misc_data_json = json.dumps(misc_data) if misc_data else None

        # Determine stealth_disadvantage and strength_req for armor
        stealth_disadvantage = False
        strength_req = None
        armor_ac = None
        armor_dex_mod = False

        if item_type in ["LA", "MA", "HA"]:
            armor_ac = item_data.get("ac")
            if item_data.get("stealth") == True:
                stealth_disadvantage = True
            if item_data.get("strength"):
                try:
                    strength_req = int(item_data["strength"])
                except ValueError:
                    pass # Handle cases where strength is not a simple integer

            # Determine armor_dex_mod
            if item_type == "LA":
                armor_dex_mod = True
            elif item_type == "MA":
                armor_dex_mod = True # Medium armor typically adds Dex mod up to +2
            elif item_type == "HA":
                armor_dex_mod = False # Heavy armor does not add Dex mod

        # Add source to the sources dictionary if not already present
        if source not in sources:
            sources[source] = current_source_id
            source_inserts.append(f"INSERT INTO sources (name) VALUES ('{source}');")
            current_source_id += 1
        
        source_id = sources[source]

        # Construct the INSERT statement for the items table
        if db_type == "postgres":
            item_inserts.append(
                f"INSERT INTO items (name, source_id, page, type, rarity, weight, value, attunement, req_attune, tier, recharge, charges, roll_data, dmg1, dmg2, dmg_type, properties, range, stealth_disadvantage, strength_req, armor_ac, armor_dex_mod, weapon_category, ammo_type, is_container, is_staff, is_wand, is_rod, is_sentient, is_cursed, is_srd, is_basic_rules, misc_data) VALUES (" 
                f"'{name}', {source_id}, {page if page is not None else 'NULL'}, {f"'{item_type}'" if item_type is not None else "NULL"}, {f"'{rarity}'" if rarity is not None else "NULL"}, {weight if weight is not None else 'NULL'}, {value if value is not None else 'NULL'}, {str(attunement).lower()}, {f"'{req_attune}'" if req_attune is not None else "NULL"}, {f"'{tier}'" if tier is not None else "NULL"}, {f"'{recharge}'" if recharge is not None else "NULL"}, {charges if charges is not None else 'NULL'}, {f"'{roll_data_json}'::jsonb" if roll_data_json is not None else "NULL"}, {f"'{dmg1}'" if dmg1 is not None else "NULL"}, {f"'{dmg2}'" if dmg2 is not None else "NULL"}, {f"'{dmg_type}'" if dmg_type is not None else "NULL"}, {f"'{properties_json}'::jsonb" if properties_json is not None else "NULL"}, {f"'{range_val}'" if range_val is not None else "NULL"}, {str(stealth_disadvantage).lower()}, {strength_req if strength_req is not None else 'NULL'}, {armor_ac if armor_ac is not None else 'NULL'}, {str(armor_dex_mod).lower()}, {f"'{weapon_category}'" if weapon_category is not None else "NULL"}, {f"'{ammo_type}'" if ammo_type is not None else "NULL"}, {str(is_container).lower()}, {str(is_staff).lower()}, {str(is_wand).lower()}, {str(is_rod).lower()}, {str(is_sentient).lower()}, {str(is_cursed).lower()}, {str(is_srd).lower()}, {str(is_basic_rules).lower()}, {f"'{misc_data_json}'::jsonb" if misc_data_json is not None else "NULL"});"
            )
        elif db_type == "sqlite":
            item_inserts.append(
                f"INSERT INTO items (name, source_id, page, type, rarity, weight, value, attunement, req_attune, tier, recharge, charges, roll_data, dmg1, dmg2, dmg_type, properties, range, stealth_disadvantage, strength_req, armor_ac, armor_dex_mod, weapon_category, ammo_type, is_container, is_staff, is_wand, is_rod, is_sentient, is_cursed, is_srd, is_basic_rules, misc_data) VALUES (" 
                f"'{name}', {source_id}, {page if page is not None else 'NULL'}, {f"'{item_type}'" if item_type is not None else "NULL"}, {f"'{rarity}'" if rarity is not None else "NULL"}, {weight if weight is not None else 'NULL'}, {value if value is not None else 'NULL'}, {int(attunement)}, {f"'{req_attune}'" if req_attune is not None else "NULL"}, {f"'{tier}'" if tier is not None else "NULL"}, {f"'{recharge}'" if recharge is not None else "NULL"}, {charges if charges is not None else 'NULL'}, {f"'{roll_data_json}'" if roll_data_json is not None else "NULL"}, {f"'{dmg1}'" if dmg1 is not None else "NULL"}, {f"'{dmg2}'" if dmg2 is not None else "NULL"}, {f"'{dmg_type}'" if dmg_type is not None else "NULL"}, {f"'{properties_json}'" if properties_json is not None else "NULL"}, {f"'{range_val}'" if range_val is not None else "NULL"}, {int(stealth_disadvantage)}, {strength_req if strength_req is not None else 'NULL'}, {armor_ac if armor_ac is not None else 'NULL'}, {int(armor_dex_mod)}, {f"'{weapon_category}'" if weapon_category is not None else "NULL"}, {f"'{ammo_type}'" if ammo_type is not None else "NULL"}, {int(is_container)}, {int(is_staff)}, {int(is_wand)}, {int(is_rod)}, {int(is_sentient)}, {int(is_cursed)}, {int(is_srd)}, {int(is_basic_rules)}, {f"'{misc_data_json}'" if misc_data_json is not None else "NULL"});"
            )

        # For item_entries, we need the ID of the item just inserted.
        # Since we're generating a script, we'll use a placeholder or rely on
        # the order of insertion if IDs are auto-incrementing.
        # For simplicity in this script, we'll assume a sequential ID for item_entries
        # based on the order of items processed. In a real DB, you'd use RETURNING ID.
        # Here, we'll just use a counter for item_id for item_entries.
        # This is a simplification and might need adjustment for actual database import.
        item_id_placeholder = len(item_inserts) # This will act as a pseudo-ID

        entries = item_data.get("entries")
        if entries:
            item_entry_inserts.extend(parse_item_entries(item_id_placeholder, entries, db_type))

    return source_inserts, item_inserts, item_entry_inserts

def main():
    script_dir = "C:\\Projects\\dnd-class-data"
    raw_data_dir = os.path.join(script_dir, "raw-data")
    postgres_output_dir = os.path.join(script_dir, "postgres")
    sqlite_output_dir = os.path.join(script_dir, "sqlite")

    os.makedirs(postgres_output_dir, exist_ok=True)
    os.makedirs(sqlite_output_dir, exist_ok=True)

    items_json_path = os.path.join(raw_data_dir, "items.json")
    items_base_json_path = os.path.join(raw_data_dir, "items-base.json")

    all_items_data = []

    with open(items_json_path, 'r', encoding='utf-8') as f:
        data = json.load(f)
        all_items_data.extend(data.get("item", []))

    with open(items_base_json_path, 'r', encoding='utf-8') as f:
        data = json.load(f)
        all_items_data.extend(data.get("baseitem", []))

    # Generate Postgres inserts
    pg_source_inserts, pg_item_inserts, pg_item_entry_inserts = generate_inserts(all_items_data, "postgres")
    with open(os.path.join(postgres_output_dir, "insert_items_postgres.sql"), "w", encoding='utf-8') as f:
        f.write("-- Source Inserts\n")
        f.write("\n".join(pg_source_inserts))
        f.write("\n\n-- Item Inserts\n")
        f.write("\n".join(pg_item_inserts))
        f.write("\n\n-- Item Entry Inserts\n")
        f.write("\n".join(pg_item_entry_inserts))

    # Generate SQLite inserts
    sqlite_source_inserts, sqlite_item_inserts, sqlite_item_entry_inserts = generate_inserts(all_items_data, "sqlite")
    with open(os.path.join(sqlite_output_dir, "insert_items_sqlite.sql"), "w", encoding='utf-8') as f:
        f.write("-- Source Inserts\n")
        f.write("\n".join(sqlite_source_inserts))
        f.write("\n\n-- Item Inserts\n")
        f.write("\n".join(sqlite_item_inserts))
        f.write("\n\n-- Item Entry Inserts\n")
        f.write("\n".join(sqlite_item_entry_inserts))

    print("Generated insert_items_postgres.sql and insert_items_sqlite.sql")

if __name__ == "__main__":
    main()
